In [1]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
from matplotlib import  pyplot as plt
import requests
import folium
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans

%matplotlib inline

## Get Web Content
Use the requests package to get the web content, then use the beautifulsoup to parse the data

In [2]:
# get the page
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = requests.get(url)

In [3]:
# load html
soup = BeautifulSoup(page.content, "lxml")

In [4]:
tablehead = soup.find("table").find_all("th")

In [5]:
# get the data
columns = []
for i in tablehead:
    columns.append(i.text.strip())

In [6]:
tabledata = soup.find("table").find_all("tr")[1:]

## Create Data
When parse the table content, collect the data into the dataframe.Finally, group by the postcode and Borough to paste the Neighbourhood together

In [7]:
# create the dataframe
data = pd.DataFrame(columns=columns)
index = 0
for i in tabledata:
    temp = []
    for item in i.find_all("td"):
        temp.append(item.text.strip())
    if "Not assigned" not in temp:
        data = pd.concat([data, pd.DataFrame(dict(zip(columns, temp)), index=[index])])
        index += 1

In [8]:
data = data.groupby(["Postcode", "Borough"])["Neighbourhood"].apply(lambda x: ", ".join(x.tolist())).reset_index()

In [9]:
data.loc[data.Postcode == "M5A"]

Postcode           Borough              Neighbourhood
53      M5A  Downtown Toronto  Harbourfront, Regent Park

## Merge Geo Infor
Merge the geo information into the data

In [10]:
# load geo data
geodata = pd.read_csv("./Geospatial_Coordinates.csv")
geodata.rename({"Postal Code":"Postcode", "Neighbourhood":"Neighborhood"}, axis=1, inplace=True)

In [11]:
# merge the geo data
data = data.merge(geodata, how="left", on="Postcode")

In [12]:
data.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

check the dataset information, like dimentions and unique Borough

In [13]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(data['Borough'].unique()),
        data.shape[0]
    )
)

The dataframe has 10 boroughs and 102 neighborhoods.


In [14]:
data.Borough.unique()

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       'Mississauga', 'Etobicoke'], dtype=object)

As we check the Capital of Canada information [Canada - Wikipedia](https://en.wikipedia.org/wiki/Canada) and [Toronto](https://en.wikipedia.org/wiki/Toronto),  let's visualizat Toronto neighborhoods in it. The location is 43°44′30″N 79°22′24″W

In [15]:
toronto = data.loc[data.Borough.str.contains("Toronto")].copy().reset_index(drop=True)
toronto.Borough.unique()

array(['East Toronto', 'Central Toronto', 'Downtown Toronto',
       'West Toronto'], dtype=object)

In [16]:
toronto.head(2)

Postcode       Borough                 Neighbourhood   Latitude  Longitude
0      M4E  East Toronto                   The Beaches  43.676357 -79.293031
1      M4K  East Toronto  The Danforth West, Riverdale  43.679557 -79.352188

In [17]:
# create map by toronto location
latitude, longitude = 43.633, -79.367
torontomap = folium.Map(location=[43.633, -79.367], zoom_start=12)

# add markers to map
for iterm in toronto[["Latitude", "Longitude", "Neighbourhood"]].iterrows():
    lat, lng, label = iterm[1]
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(torontomap) 
torontomap

Next, we are going to start utlizing the API to explore informaton
### Define credentials

In [18]:
with open("./credentials.txt", "r") as file:
    CLIENT_ID = file.readline().strip()
    CLIENT_SECRET = file.readline().strip()

In [19]:
VERSION = '20180605'
if False:
    print('Your credentails:')
    print('CLIENT_ID: ' + CLIENT_ID)
    print('CLIENT_SECRET:' + CLIENT_SECRET)

Explore the first neighborhood in our dataframe, we can get the name and locaton

In [20]:
toronto.columns

Index(['Postcode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude'], dtype='object')

In [21]:
neighborhood_latitude = toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [22]:
# create URL
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

Right now, let's get the top 100 venues that are in current location within a radius of 500 meters. First, let's create the GET request URL. Name your URL url.Send the GET request and examine the resutls

In [23]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c951ebdf594df6af0e9a818'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ad4c062f964a52011f820e3',
       'name': 'The Big Carrot Natural Food Market',
       'location': {'address': '125 Southwood Dr',
        'lat': 43.678879,
        'lng': -79.297734,
        'labeledLatLngs': [{'label': 'display',
          'lat

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab.

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [25]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
1                 Grover Pub and Grub                Pub  43.679181 -79.297215
2                           Starbucks        Coffee Shop  43.678798 -79.298045
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [26]:
print('{} venues were returned by Foursquare At the current location.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare At the current location.


### Explore neighbourhoods in toronto


In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, verbose=False):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        if verbose:
            print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [28]:
toronto_venues = getNearbyVenues(names=toronto['Neighbourhood'],
                                 latitudes=toronto['Latitude'],
                                 longitudes=toronto['Longitude'],
                                 verbose=True
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

## Analyze Neighbourhood
Next, we check the dataframe venue information

In [29]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant      ...        Toy / Game Store  \
0                 0                    0      ...                       0   
1                 0                    0      ...                       0   
2                 0                    0      ...                       0   
3                 0                    0      ...                       0   
4                 0                    0      ...                       0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                      0         0          0            0              0  
1                      0         0          0            0              0  
2                      0         0          0            0              0  
3                      0         0          0            0              0  
4                      0         0          0            0              0  

[5 rows x 236 columns]

In [30]:
toronto_onehot.shape

(1694, 236)

There are 236 Category information. Next we group rows by neighbourhood and  extract mean

In [31]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                           Adelaide, King, Richmond          0.0   
1                                        Berczy Park          0.0   
2       Brockton, Exhibition Place, Parkdale Village          0.0   
3  Business Reply Mail Processing Centre 969 Eastern          0.0   
4  CN Tower, Bathurst Quay, Island airport, Harbo...          0.0   

   Accessories Store  Afghan Restaurant   Airport  Airport Food Court  \
0               0.01                0.0  0.000000            0.000000   
1               0.00                0.0  0.000000            0.000000   
2               0.00                0.0  0.000000            0.000000   
3               0.00                0.0  0.000000            0.000000   
4               0.00                0.0  0.076923            0.076923   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0      0.000000        0.000000         0.000000          0.000000   
1      0.000000        0.000000         0.000000          0.000000   
2      0.000000        0.000000         0.000000          0.000000   
3      0.000000        0.000000         0.000000          0.000000   
4      0.076923        0.153846         0.153846          0.153846   

       ...        Toy / Game Store  Trail  Train Station  \
0      ...                     0.0    0.0            0.0   
1      ...                     0.0    0.0            0.0   
2      ...                     0.0    0.0            0.0   
3      ...                     0.0    0.0            0.0   
4      ...                     0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                           0.01               0.0                    0.0   
1                           0.00               0.0                    0.0   
2                           0.00               0.0                    0.0   
3                           0.00               0.0                    0.0   
4                           0.00               0.0                    0.0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0      0.01        0.0          0.0           0.01  
1      0.00        0.0          0.0           0.00  
2      0.00        0.0          0.0           0.00  
3      0.00        0.0          0.0           0.00  
4      0.00        0.0          0.0           0.00  

[5 rows x 236 columns]

In [32]:
toronto_grouped.shape

(38, 236)

Explore the top 5 common category in each venues. Then sort the venuses in descending order and store information

In [33]:
top_venue_num = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(top_venue_num))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.06
1              Bar  0.04
2             Café  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.08
1        Cocktail Bar  0.05
2          Restaurant  0.05
3  Seafood Restaurant  0.03
4              Bakery  0.03


----Brockton, Exhibition Place, Parkdale Village----
               venue  freq
0     Breakfast Spot  0.10
1               Café  0.10
2        Coffee Shop  0.10
3  Convenience Store  0.05
4         Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1                 Spa  0.06
2       Auto Workshop  0.06
3                Park  0.06
4          Comic Shop  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.15
1  Airport Te

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [35]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Bar       Thai Restaurant                  Café   
1            Restaurant          Cocktail Bar                   Pub   
2           Coffee Shop        Breakfast Spot                   Gym   
3        Farmers Market                Garden            Skate Park   
4       Airport Service      Airport Terminal         Boat or Ferry   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                 Hotel   American Restaurant   
1        Farmers Market           Cheese Shop    Seafood Restaurant   
2          Climbing Gym    Falafel Restaurant     Convenience Store   
3               Brewery            Comic Shop  Fast Food Restaurant   
4       Harbor / Marina              Boutique               Airport   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery      Sushi Restaurant       Asian Restaurant  
1    Italian Restaurant                Bakery             Steakhouse  
2         Burrito Place               Stadium                    Bar  
3                   Spa         Burrito Place                Butcher  
4    Airport Food Court          Airport Gate       Sculpture Garden

## Cluster Neighbourhoods
Run k_means to cluster the neighbourhood into 4 clusters. Then create dataframe that includes as well as the top 10 stores for each neighbourhood

In [36]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [37]:
neighborhoods_venues_sorted.columns

Index(['Neighborhood', '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue', '6th Most Common Venue',
       '7th Most Common Venue', '8th Most Common Venue',
       '9th Most Common Venue', '10th Most Common Venue'],
      dtype='object')

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [39]:
neighborhoods_venues_sorted.head(2)#set_index('Neighborhood')

Cluster Labels              Neighborhood 1st Most Common Venue  \
0               1  Adelaide, King, Richmond           Coffee Shop   
1               1               Berczy Park           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Bar       Thai Restaurant                  Café   
1            Restaurant          Cocktail Bar                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                 Hotel   American Restaurant   
1        Farmers Market           Cheese Shop    Seafood Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery      Sushi Restaurant       Asian Restaurant  
1    Italian Restaurant                Bakery             Steakhouse

In [42]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto.merge(neighborhoods_venues_sorted, left_on='Neighbourhood', right_on="Neighborhood")

toronto_merged.head() # check the last columns!

Postcode          Borough                   Neighbourhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3      M4M     East Toronto                 Studio District  43.659526   
4      M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels                    Neighborhood  \
0 -79.293031               0                     The Beaches   
1 -79.352188               1    The Danforth West, Riverdale   
2 -79.315572               1  The Beaches West, India Bazaar   
3 -79.340923               1                 Studio District   
4 -79.388790               2                   Lawrence Park   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     Health Food Store           Coffee Shop                   Pub   
1      Greek Restaurant           Coffee Shop        Ice Cream Shop   
2         Movie Theater          Intersection        Ice Cream Shop   
3                  Café           Coffee Shop    Italian Restaurant   
4                  Park           Swim School              Bus Line   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Diner  Fast Food Restaurant        Farmers Market   
1    Italian Restaurant             Bookstore           Yoga Studio   
2                   Pub        Sandwich Place     Fish & Chips Shop   
3                Bakery             Gastropub   American Restaurant   
4         Women's Store        Discount Store        Farmers Market   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0    Falafel Restaurant           Event Space  Ethiopian Restaurant   
1               Brewery       Bubble Tea Shop                  Café   
2      Sushi Restaurant            Steakhouse  Fast Food Restaurant   
3           Fish Market             Juice Bar             Bookstore   
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

      10th Most Common Venue  
0          Electronics Store  
1                 Restaurant  
2                Coffee Shop  
3  Latin American Restaurant  
4          Electronics Store

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters
Examine each cluster and determin the discriminating venue categories aht distinguish each cluster

### Cluster 1

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  \
0     East Toronto               0   
9  Central Toronto               0   

                                        Neighborhood 1st Most Common Venue  \
0                                        The Beaches     Health Food Store   
9  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop                   Pub                 Diner   
9                   Pub   American Restaurant    Light Rail Station   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
9            Sports Bar        Medical Center           Supermarket   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Event Space  Ethiopian Restaurant      Electronics Store  
9      Sushi Restaurant            Bagel Shop    Fried Chicken Joint

### Cluster 2

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  \
1       East Toronto               1   
2       East Toronto               1   
3       East Toronto               1   
5    Central Toronto               1   
6    Central Toronto               1   
7    Central Toronto               1   
11  Downtown Toronto               1   
12  Downtown Toronto               1   
13  Downtown Toronto               1   
14  Downtown Toronto               1   
15  Downtown Toronto               1   
16  Downtown Toronto               1   
17  Downtown Toronto               1   
18  Downtown Toronto               1   
19  Downtown Toronto               1   
20  Downtown Toronto               1   
21  Downtown Toronto               1   
24   Central Toronto               1   
25  Downtown Toronto               1   
26  Downtown Toronto               1   
27  Downtown Toronto               1   
28  Downtown Toronto               1   
29  Downtown Toronto               1   
30  Downtown Toronto               1   
31      West Toronto               1   
32      West Toronto               1   
33      West Toronto               1   
34      West Toronto               1   
35      West Toronto               1   
36      West Toronto               1   
37      East Toronto               1   

                                         Neighborhood 1st Most Common Venue  \
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      The Beaches West, India Bazaar         Movie Theater   
3                                     Studio District                  Café   
5                                    Davisville North                 Hotel   
6                                  North Toronto West   Sporting Goods Shop   
7                                          Davisville           Pizza Place   
11                        Cabbagetown, St. James Town           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                          Harbourfront, Regent Park           Coffee Shop   
14                           Ryerson, Garden District           Coffee Shop   
15                                     St. James Town           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
18                           Adelaide, King, Richmond           Coffee Shop   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20           Design Exchange, Toronto Dominion Centre           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
24                The Annex, North Midtown, Yorkville                  Café   
25                     Harbord, University of Toronto                  Café   
26          Chinatown, Grange Park, Kensington Market                  Café   
27  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
29             First Canadian Place, Underground city           Coffee Shop   
30                                           Christie         Grocery Store   
31                       Dovercourt Village, Dufferin              Pharmacy   
32                           Little Portugal, Trinity                   Bar   
33       Brockton, Exhibition Place, Parkdale Village                  Café   
34                      High Park, The Junction South    Mexican Restaurant   
35                             Parkdale, Roncesvalles        Breakfast Spot   
36                                 Runnymede, Swansea           Pizza Place   
37  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   

    2nd Most Common Venue          3rd Most Common Venue  \
1             Coffee Shop                 Ice Cream Shop   
2            Intersection                 Ice Cream Shop   
3             Coffee Sh

### Cluster 3

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels   Neighborhood 1st Most Common Venue  \
4  Central Toronto               2  Lawrence Park                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4           Swim School              Bus Line         Women's Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4        Discount Store        Farmers Market    Falafel Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4           Event Space  Ethiopian Restaurant      Electronics Store

### Cluster 4

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels                 Neighborhood  \
8  Central Toronto               3  Moore Park, Summerhill East   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
8            Playground          Tennis Court         Women's Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
8    Dim Sum Restaurant        Farmers Market    Falafel Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
8           Event Space  Ethiopian Restaurant     Electronics Store   

        10th Most Common Venue  
8  Eastern European Restaurant